modules/gdprEnforcement.js

```diff
@@ -0,0 +1,181 @@
+/**
+ * This module gives publishers extra set of features to enforce individual purposes of TCF v2
+ */
+
+import * as utils from '../src/utils.js';
+import { config } from '../src/config.js';
+import { hasDeviceAccess } from '../src/utils.js';
+import adapterManager, { gdprDataHandler } from '../src/adapterManager.js';
+import find from 'core-js/library/fn/array/find.js';
+import includes from 'core-js/library/fn/array/includes.js';
+import { registerSyncInner } from '../src/adapters/bidderFactory.js';
+import { getHook } from '../src/hook.js';
+import { validateStorageEnforcement } from '../src/storageManager.js';
+
+const purpose1 = 'storage';
+
+let addedDeviceAccessHook = false;
+let enforcementRules;
+
+function getGvlid() {
+  let gvlid;
+  const bidderCode = config.getCurrentBidder();
+  if (bidderCode) {
+    const bidder = adapterManager.getBidAdapter(bidderCode);
+    gvlid = bidder.getSpec().gvlid;
+  } else {
+    utils.logWarn('Current bidder not found');
+  }
+  return gvlid;
+}
+
+/**
+ * This function takes in rules and consentData as input and validates against the consentData provided. If it returns true Prebid will allow the next call else it will log a warning
+ * @param {Object} rules enforcement rules set in config
+ * @param {Object} consentData gdpr consent data
+ * @returns {boolean}
+ */
+function validateRules(rule, consentData, currentModule, gvlid) {
+  let isAllowed = false;
+  if (rule.enforcePurpose && rule.enforceVendor) {
+    if (includes(rule.vendorExceptions, currentModule) ||
+        (utils.deepAccess(consentData, 'vendorData.purpose.consents.1') === true && utils.deepAccess(consentData, `vendorData.vendor.consents.${gvlid}`) === true)) {
+      isAllowed = true;
+    }
+  } else if (rule.enforcePurpose === false && rule.enforceVendor === true) {
+    if (includes(rule.vendorExceptions, currentModule) ||
+        (utils.deepAccess(consentData, `vendorData.vendor.consents.${gvlid}`) === true)) {
+      isAllowed = true;
+    }
+  } else if (rule.enforcePurpose === false && rule.enforceVendor === false) {
+    if ((includes(rule.vendorExceptions, currentModule) &&
+        (utils.deepAccess(consentData, 'vendorData.purpose.consents.1') === true && utils.deepAccess(consentData, `vendorData.vendor.consents${gvlid}`) === true)) ||
```

Comment:
Similarly, vendorExceptions wasn't intended affect the high level enforcement of the purpose. So I think this should be interpreted as "only consider the user's vendor choices for the exceptions".

(includes(rule.vendorExceptions, currentModule) && utils.deepAccess(consentData, `vendorData.vendor.consents${gvlid}`) === true